department-of-veterans-affairs/caseflow
# Log, Dont Send, 7 & 3 Day Reminder Emails for Video/Central Hearings

Resolves [CASEFLOW-1519](https://vajira.max.gov/browse/CASEFLOW-1519)

### Description
This PR:
- Starts sending alerts to Rails.logger (Datadog) when we would send a 7 day or 3 day reminder for a Video or Central hearing.
- Does not start sending 7 day or 3 day reminders for Video or Central hearings.
- Does not add a row in the Email Notification History section of Hearing Details, as we are not sending emails yet.

### Acceptance Criteria From Jira
- [ ] When an appellant or representative email are entered for Video or Central hearings, then reminder emails are sent to those addresses 7 days before the hearing and again 2/3 days before the hearing 
- [ ] Record to DataDog when email would be sent for Video/Central hearings (but don't send until a later update)
- [ ] Sent emails each appear as a row in the Email Notification History section of Hearing Details

### Later: Actually Send 7 & 3 day Reminder Emails for Video/Central
At some point we will want to actually send the emails for 7 & 3 day reminders. To do that, we'll make a PR with these changes.

```ruby
# app/services/hearings/reminder_service.rb:23
which_type_of_reminder_to_send # Remove the "&& hearing.virtual?" as well as the comment above

# app/jobs/hearings/send_email.rb:25
# return if !hearing.virtual_hearing <- Remove that, and the comment above

# spec/jobs/hearings/send_reminder_emails_job_spec.rb:372
# Follow the directions in the comment to get tests passing

# spec/services/hearings/reminder_service_spec.rb:170
# Uncomment the tests, they should pass
```
### High Level Code Change Overview
- `HearingRepository::maybe_ready_for_reminder_email` changed to query the db for `hearing`s rather than `virtual_hearing`s.
- `SendReminderEmailsJob` changed to expect hearings from `maybe_ready_for_reminder_email` and call `ReminderService` differently when hearings are Video/Central vs Virtual. Also changed to send a `hearing` to `SendEmail`.
- `ReminderService` changed to work with `hearing` rather than `virtual_hearing` so it can handle Video/Central hearings.
- `SendEmail` changed to accept either a `hearing` or `virtual_hearing`, but always work with the `hearing` so it can send emails for Video/Central hearings.

### Query Timing Notes
I was discussing this with Rubaiyat and since I'm changing some db access in `maybe_ready_for_reminder_email` it seemed like a good idea to figure out if that was going to cause slowness in prod. Since this is an access only method I was able to run three different versions of `maybe_ready_for_reminder_email` in prod:

```ruby
# This method is the one in master branch
old_timing = Benchmark.measure { old_method }
=> #<Benchmark::Tms:0x00007fb92301dd60 @label="", @real=6.155024803010747, @cstime=0.0, @cutime=0.0, @stime=0.007949999999999902,...
old_method.count
=> 565

# This method is the version in [this commit](https://github.com/department-of-veterans-affairs/caseflow/pull/16623/commits/2c20acd60a18250cace42f033f2d6c3a1c7987be), it does not exclude travel days
new_timing_no_filter = Benchmark.measure { new_method_no_filter }
=> #<Benchmark::Tms:0x00007fb9253395d8 @label="", @real=0.8160953350015916, @cstime=0.0, @cutime=0.0, @stime=0.011797000000000057...
new_method_no_filter.count
=> 732

# This method is the version in [this commit](https://github.com/department-of-veterans-affairs/caseflow/pull/16623/commits/20af48dbe4a1ae69bc316f64cf3bf6de6311699d), it excludes travel days
new_timing_filter = Benchmark.measure { new_method_with_filter }
=> #<Benchmark::Tms:0x00007fb925ff4a70 @label="", @real=0.5795998780085938, @cstime=0.0, @cutime=0.0, @stime=0.007954000000000017...
new_method_with_filter.count
=> 732
```

### Testing Plan
I think that the best way to test this functionality is through the Rspec tests. Setting up the data necessary any other way seems complex. Something like this should help examine the changes:
- Reading the tests in the three changed spec files.
- Add `binding.pry` as the first line of `SendReminderEmailsJob:send_reminder_emails`.
- Run: `make one-test spec/services/hearings/reminder_service_spec.rb`
- `step` and `next` through the code, making sure it behaves appropriately.
